src/Roslyn.Diagnostics.Analyzers/Core/ExportedPartsShouldHaveImportingConstructor.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                var exportAttributeV1 = compilationContext.Compilation.GetTypeByMetadataName("System.ComponentModel.Composition.ExportAttribute");
+                var importingConstructorAttributeV1 = compilationContext.Compilation.GetTypeByMetadataName("System.ComponentModel.Composition.ImportingConstructorAttribute");
+                var exportAttributeV2 = compilationContext.Compilation.GetTypeByMetadataName("System.Composition.ExportAttribute");
+                var importingConstructorAttributeV2 = compilationContext.Compilation.GetTypeByMetadataName("System.Composition.ImportingConstructorAttribute");
+
+                if (exportAttributeV1 is null && exportAttributeV2 is null)
+                {
+                    // We don't need to check assemblies unless they're referencing MEF, so we're done
+                    return;
+                }
+
+                compilationContext.RegisterSymbolAction(symbolContext =>
+                {
+                    var namedType = (INamedTypeSymbol)symbolContext.Symbol;
+                    var namedTypeAttributes = namedType.GetApplicableAttributes();
+
+                    AnalyzeSymbolForAttribute(ref symbolContext, exportAttributeV1, importingConstructorAttributeV1, namedType, namedTypeAttributes);
+                    AnalyzeSymbolForAttribute(ref symbolContext, exportAttributeV2, importingConstructorAttributeV2, namedType, namedTypeAttributes);
+                }, SymbolKind.NamedType);
+            });
+        }
+
+        private static void AnalyzeSymbolForAttribute(ref SymbolAnalysisContext context, INamedTypeSymbol exportAttributeOpt, INamedTypeSymbol importingConstructorAttribute, INamedTypeSymbol namedType, IEnumerable<AttributeData> namedTypeAttributes)
+        {
+            if (exportAttributeOpt is null)
+            {
+                return;
+            }
+
+            var exportAttributeApplication = namedTypeAttributes.FirstOrDefault(ad => ad.AttributeClass.DerivesFrom(exportAttributeOpt));
+            if (exportAttributeApplication is null)
+            {
+                return;
+            }
+
+            IMethodSymbol importingConstructor = null;
+            ImmutableArray<IMethodSymbol> nonImportingConstructors = ImmutableArray<IMethodSymbol>.Empty;
+            foreach (var constructor in namedType.Constructors)
+            {
+                if (constructor.IsImplicitlyDeclared)
+                {
+                    // '{0}' is MEF-exported and should have a single importing constructor of the correct form
+                    context.ReportDiagnostic(Diagnostic.Create(Rule, exportAttributeApplication.ApplicationSyntaxReference.GetSyntax().GetLocation(), ScenarioProperties.ImplicitConstructor, namedType.Name));
```

Comment:
`GetApplicableAttributes` returns attributes from the base types as well, which may be in metadata and can have a null `ApplicationSyntaxReference`. For example, see https://github.com/dotnet/roslyn-analyzers/blob/32d8f1e397439035f0ecb5f61a9e672225f0ecdb/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerAttributeAnalyzer.cs#L83

Reply:
Adding a null check